File: SDK/Agency.cs
using AutoMapper;
using Agience.SDK.Mappings;
using Microsoft.Extensions.Logging;
using System.Text.Json;
using Agience.SDK.Models.Messages;
using Microsoft.SemanticKernel.ChatCompletion;
using Microsoft.SemanticKernel;

namespace Agience.SDK
{
    [AutoMap(typeof(Models.Entities.Agency), ReverseMap = true)]
    public class Agency : Models.Entities.Agency
    {
        // TODO: Agent Orchestration via Agency. One such method: ACE Layer processing
        // https://github.com/daveshap/ACE_Framework/blob/main/publications/Conceptual%20Framework%20for%20Autonomous%20Cognitive%20Entities%20(ACE).pdf
        // https://github.com/daveshap/ACE_Framework/blob/main/ACE_PRIME/HelloAF/src/ace/resources/core/hello_layers/prompts/templates/ace_context.md

        public event Func<IEnumerable<ChatMessageContent>, Task>? HistoryUpdated;

        public bool IsConnected { get; private set; }
        internal string? RepresentativeId { get; private set; }
        public string Timestamp => _broker.Timestamp;

        private readonly Authority _authority;
        private readonly Broker _broker;
        private readonly ILogger<Agency> _logger;
        private readonly IMapper _mapper;
        
        private readonly Dictionary<string, DateTime> _agentJoinTimestamps = new();
        private readonly Dictionary<string, Agent> _agents = new();
        private readonly List<string> _localAgentIds = new();        

        private readonly ChatHistory _history = new(); // TODO: History should be persistent.
        //private readonly History _history; // TODO: Use History instead of ChatHistory

        internal Agency(Authority authority, Broker broker, ILogger<Agency> logger)
        {
            _authority = authority;
            _broker = broker;
            _logger = logger;
            _mapper = AutoMapperConfig.GetMapper();
            //_history = new History(null, Id); // TODO: use idProvider to get a unique id
        }

        internal async Task Connect()
        {
            if (!IsConnected)
            {
                await _broker.Subscribe(_authority.AgencyTopic("+", Id!), _broker_ReceiveMessage);
                IsConnected = true;
            }
        }

        internal async Task Disconnect()
        {
            if (IsConnected)
            {
                await _broker.Unsubscribe(_authority.AgencyTopic("+", Id!));
                IsConnected = false;
            }
        }

        private Task _broker_ReceiveMessage(BrokerMessage message)
        {
            if (message.SenderId == null) { return Task.CompletedTask; }

            // Incoming Agent Join message
            if (message.Type == BrokerMessageType.EVENT &&
                message.Data?["type"] == "join" &&
                message.Data?["agent"] != null &&
                message.Data?["timestamp"] != null)
            {
                var timestamp = DateTime.TryParse(message.Data?["timestamp"], out DateTime result) ? (DateTime?)result : null;
                var agentId = message.Data?["agent_id"];

                if (agentId == message.SenderId && timestamp != null)
                {
                    ReceiveJoin(agentId, (DateTime)timestamp);
                }
            }

            // Incoming Representative Claim message
            if (message.Type == BrokerMessageType.EVENT &&
                message.Data?["type"] == "representative_claim" &&
                message.Data?["agent_id"] != null &&
                message.Data?["timestamp"] != null)
            {
                var timestamp = DateTime.TryParse(message.Data?["timestamp"], out DateTime result) ? (DateTime?)result : null;
                var agentId = message.Data?["agent_id"];

                if (agentId == message.SenderId && timestamp != null)
                {
                    ReceiveRepresentativeClaim(agentId, timestamp);
                }
            }
            return Task.CompletedTask;
        }

        private void ReceiveJoin(string agentId, DateTime timestamp)
        {
            _logger.LogInformation($"ReceiveJoin {_agents[agentId].Name}");

            _agentJoinTimestamps[agentId] = timestamp;            

            if (RepresentativeId != null && _localAgentIds.Contains(RepresentativeId))
            {
                //SendAgentWelcome(agentId);
            }
        }

        /*
        private void SendAgentWelcome(string agentId)
        {
            _logger.LogInformation($"SendAgentWelcome to {_agents[agentId].Name} with {_agents.Values.Count} Agents.");

            _broker.Publish(new BrokerMessage()
            {
                Type = BrokerMessageType.EVENT,
                Topic = _authority.AgentTopic(Id!, agent.Id!),
                Data = new Data
                {
                    { "type", "welcome" },
                    { "timestamp", _broker.Timestamp },
                    { "agency", JsonSerializer.Serialize(_mapper.Map<Models.Entities.Agency>(this)) },
                    { "representative_id", RepresentativeId },
                    //{ "agents", JsonSerializer.Serialize(_agents.Values.Select(a => a.Item1).ToList()) },
                    //{ "agent_timestamps", JsonSerializer.Serialize(_agents.ToDictionary(a => a.Key, a => a.Value.Item2)) }
                }
            });
        }*/

        /*
        internal void ReceiveWelcome(Models.Entities.Agency agency,
                                     string representativeId,
                                     List<Models.Entities.Agent> agents,
                                     Dictionary<string, DateTime> agentTimestamps,                                            
                                     DateTime timestamp)
        {
            _logger?.LogInformation($"ReceiveWelcome from {agency.Name} {GetAgentName(representativeId)}");

            if (RepresentativeId != representativeId)
            {
                RepresentativeId = representativeId;
                _logger?.LogInformation($"Set Representative {GetAgentName(RepresentativeId)}");
            }

            foreach (var agent in agents)
            {
                _agents[agent.Id!] = (agent, agentTimestamps[agent.Id!]);
            }
        }
        */

        // TODO: Handle race conditions
        // Network Latency, Simultaneous Joins, etc.
        private void ReceiveRepresentativeClaim(string agentId, DateTime? timestamp)
        {
            _logger.LogInformation($"ReceiveRepresentativeClaim from {_agents[agentId].Name}");

            if (RepresentativeId != agentId)
            {
                RepresentativeId = agentId;
                _logger?.LogInformation($"Set Representative {GetAgentName(RepresentativeId)}");
            }

            /*
            if (_localAgentIds.Contains(RepresentativeId))
            {
                var repJoinTime = _agentJoinTimestamps[RepresentativeId];
                foreach (var agent in _agentJoinTimestamps.Where(a => a.Value >= repJoinTime))
                {
                    SendAgentWelcome(agent.Value);
                }
            }*/
        }

        internal string? GetAgentName(string agentId)
        {
            return _agents.TryGetValue(agentId, out Agent? agent) ? agent.Name : null;
        }

        internal void AddLocalAgent(Agent agent)
        {
            _agents[agent.Id] = agent;
            _localAgentIds.Add(agent.Id);
        }

        internal Agent? GetLocalAgent(string agentId)
        {
            return _localAgentIds.Contains(agentId) ? _agents[agentId] : null;
        }
                
        public async Task InformAsync(string message)
        {
            var chatMessage = new ChatMessageContent(AuthorRole.User, message);
            _history.Add(chatMessage);
            HistoryUpdated?.Invoke([chatMessage]);
        }

        public async Task<IEnumerable<ChatMessageContent>> GetHistory()
        {
            return _history;
        }
    }
}

File: python/src/core/agency.py
import asyncio
import logging
from typing import Callable, Dict, List, Optional
from datetime import datetime
from dataclasses import dataclass, field


from models.messages.broker_message import BrokerMessage, BrokerMessageType

logger = logging.getLogger(__name__)


@dataclass
class Agency:
    id: str
    name: str
    authority: "Authority"
    broker: "Broker"
    logger: logging.Logger = field(default_factory=logging.getLogger)
    chat_message_received: Optional[Callable[[dict], None]] = None
    representative_id: Optional[str] = None
    is_connected: bool = False

    # Internal data stores
    agent_join_timestamps: Dict[str, datetime] = field(default_factory=dict)
    agents: Dict[str, "Agent"] = field(default_factory=dict)
    local_agent_ids: List[str] = field(default_factory=list)
    chat_history: List[dict] = field(default_factory=list)

    async def connect(self):
        """Connects the agency to the MQTT broker."""
        if not self.is_connected:
            topic = self.authority.agency_topic("+", self.id)
            await self.broker.subscribe(topic, self._broker_receive_message)
            self.is_connected = True
            logger.info(f"Agency {self.name} connected to broker.")

    async def disconnect(self):
        """Disconnects the agency from the MQTT broker."""
        if self.is_connected:
            topic = self.authority.agency_topic("+", self.id)
            await self.broker.unsubscribe(topic)
            self.is_connected = False
            logger.info(f"Agency {self.name} disconnected from broker.")

    async def _broker_receive_message(self, message: BrokerMessage):
        """Handles messages received from the broker."""
        if message.sender_id is None:
            return

        # Handle "join" messages
        if (
            message.type == BrokerMessageType.EVENT
            and message.data.get("type") == "join"
            and message.data.get("agent_id") is not None
            and message.data.get("timestamp") is not None
        ):
            timestamp = self._parse_timestamp(message.data["timestamp"])
            agent_id = message.data["agent_id"]
            if agent_id == message.sender_id and timestamp:
                self._receive_join(agent_id, timestamp)

        # Handle "representative_claim" messages
        elif (
            message.type == BrokerMessageType.EVENT
            and message.data.get("type") == "representative_claim"
            and message.data.get("agent_id") is not None
            and message.data.get("timestamp") is not None
        ):
            timestamp = self._parse_timestamp(message.data["timestamp"])
            agent_id = message.data["agent_id"]
            if agent_id == message.sender_id and timestamp:
                self._receive_representative_claim(agent_id, timestamp)

    def _parse_timestamp(self, timestamp: str) -> Optional[datetime]:
        """Parses a timestamp string into a datetime object."""
        try:
            return datetime.fromisoformat(timestamp)
        except ValueError:
            logger.error(f"Invalid timestamp format: {timestamp}")
            return None

    def _receive_join(self, agent_id: str, timestamp: datetime):
        """Handles an agent joining the agency."""
        logger.info(f"Agent {agent_id} joined agency {self.name}.")
        self.agent_join_timestamps[agent_id] = timestamp
        if self.representative_id and self.representative_id in self.local_agent_ids:
            self._send_agent_welcome(agent_id)

    def _send_agent_welcome(self, agent_id: str):
        """Sends a welcome message to a newly joined agent."""
        agent = self.agents.get(agent_id)
        if agent:
            logger.info(f"Sending welcome message to agent {agent.name}.")
            self.broker.publish(
                BrokerMessage(
                    type=BrokerMessageType.EVENT,
                    topic=self.authority.agent_topic(self.id, agent.id),
                    data={
                        "type": "welcome",
                        "timestamp": self.broker.timestamp,
                        "agency": {"id": self.id, "name": self.name},
                        "representative_id": self.representative_id,
                    },
                )
            )

    def _receive_representative_claim(self, agent_id: str, timestamp: datetime):
        """Handles a representative claim event."""
        logger.info(f"Agent {agent_id} claimed representative role for agency {self.name}.")
        if self.representative_id != agent_id:
            self.representative_id = agent_id
            logger.info(f"Set representative to agent {agent_id}.")

    def add_local_agent(self, agent: "Agent"):
        """Adds a local agent to the agency."""
        self.agents[agent.id] = agent
        self.local_agent_ids.append(agent.id)

    def get_local_agent(self, agent_id: str) -> Optional["Agent"]:
        """Retrieves a local agent by ID."""
        return self.agents.get(agent_id) if agent_id in self.local_agent_ids else None

    async def inform_async(self, message: str):
        """Sends an informational message to the chat."""
        chat_message = {"role": "user", "message": message}
        self.chat_history.append(chat_message)
        if self.chat_message_received:
            self.chat_message_received(
                {"agency_id": self.id, "message": chat_message}
            )

    async def get_history(self) -> List[dict]:
        """Retrieves the agency's chat history."""
        return self.chat_history
